tools/src/code-review/index.ts

```diff
@@ -0,0 +1,139 @@
+  const previousReviews = (await GitHub.listPullRequestReviewsAsync(pr.number)).filter(
+    (review) => review.user?.login === user.login
+  );
+
+  // Generate review body and decide whether the review needs to request for changes or not.
+  const event = getReviewEventFromOutputs(outputs);
+  const comments = getReviewCommentsFromOutputs(outputs);
+  const body = generateReviewBodyFromOutputs(activeOutputs, comments.length > 0, pr.head.sha);
+
+  // If it's the first review and there is nothing to complain,
+  // just return early — no need to bother PR's author.
+  if (!activeOutputs.length && !comments.length && !previousReviews.length) {
+    logger.success('🥳 Everything looks good to me! There is no need to submit a review.');
+    return;
+  }
+
+  // Reset my reviews' current state if I previously requested for changes.
+  if (previousReviews[previousReviews.length - 1]?.state) {
+    logger.info('🙈 Resetting my review state by re-requesting');
+    await GitHub.requestPullRequestReviewersAsync(pr.number, [user.login]);
+  }
+
+  // Create new pull request review.
+  const review = await GitHub.createPullRequestReviewAsync(pr.number, {
+    body,
+    event,
+    comments,
+  });
+  logger.info('📝 Submitted new review at:', chalk.blue(review.html_url));
+
+  // Previous reviews are no longer useful — we would delete them, but
+  // unfortunately they cannot be deleted entirely so we only make them smaller.
+  await invalidatePreviousReviewsAsync(pr.number, previousReviews, review);
+
+  logger.success("🥳 I'm done!");
+}
+
+/**
+ * Marks previous reviews as outdated by changing its body and linking to the latest one.
+ * Probably no need to keep the old body for history as GitHub shows previous revisions of edited comments.
+ */
+async function invalidatePreviousReviewsAsync(
+  prNumber: number,
+  previousReviews: GitHub.PullRequestReview[],
+  newReview: GitHub.PullRequestReview
+): Promise<void> {
+  for (const review of previousReviews) {
+    await GitHub.updatePullRequestReviewAsync(
+      prNumber,
+      review.id,
+      `*The review previously left here is no longer valid, jump to the latest one 👉 ${newReview.html_url}*`
+    );
+  }
+  if (previousReviews.length > 0) {
+    logger.info('💥 Invalidated previous reviews');
+
+    // In order not to exceed rate limits, it should be enough to remove comments only from the last review.
+    await GitHub.deleteAllPullRequestReviewCommentsAsync(
+      prNumber,
+      previousReviews[previousReviews.length - 1].id
+    );
+  }
+}
+
+/**
+ * If any of the check failed, we want the review to request for changes.
+ * Otherwise, it's just a comment (and so fixes are not obligatory).
+ * There is no case where we approve the PR — we still want a human to review these changes :)
+ */
+function getReviewEventFromOutputs(outputs: ReviewOutput[]): GitHub.PullRequestReviewEvent {
+  return outputs.some((output) => output.status >= ReviewStatus.ERROR)
```

Comment:
```suggestion
  return outputs.some((output) => output.status === ReviewStatus.ERROR)
```
same reasoning as other comment, unless we expect there to be more review statuses?